### DISCUSSION TITLE
LaTeX newcommands across jupyter notebook cells?

### DISCUSSION THREAD

**Discussion by realfogman** [sagemathinc/cocalc#6177] [Q&A]
Hello, I'm not sure if this question is answered at someplace else, but I couldn't find a proper solution.
Is there a way to define multiple LaTeX \newcommands  and use them across multiple jupyter notebook cells (one document)?
Maybe a config file or something, which can be included or edited (or maybe LaTeX \include, \input command somehow) ?

In jupyter notebook and juypter lab, defining a \newcommand and using it across cells this is possible, although it will leave you with problems in the converting to LaTeX-pdf process.

**Comment by mforbes** [comment]
If I recall correctly, this might be an issue with using KaTeX (much faster) vs MathJaX (supports definitions across cells).  Have you tried disabling KaTeX to see if that fixes things?  Still would be best if both support this.

**Update:** I did some playing and it seems that both KaTeX and MathJaX properly support definitions across cells however:

1. MathJaX is much slower, and can serious rendering issues - even scrolling away then scrolling back can cause jerking as the page refreshes:

  ![CoCalcSlowMathJaX](https://user-images.githubusercontent.com/1028433/196002318-39d7c73d-a00f-4bcd-9c1a-67746e587604.gif)

   *This was rendering [this page](https://cocalc.com/projects/d8321832-12e5-4c52-ac0b-46b5f29e9740/files/MathBug.ipynb#id=7cd1d3) with Chrome 106.0.5249.103 (Official Build) (arm64) with KaTeX rendering turned off in my account settings, but is similar with Safari.*

2. Note that changing a definition in a previous cell will not apply to other cells until they are re-rendered.  Saving and reloading the page works.  *(Is there some way to detect that a $\newcommand$ has been edited and re-render the whole document?  It might be nice feature.)*
2. The browser cache can cause problems: learn to hard refresh it by doing a hard reload (⌘⇧R for Chrome on a Mac).
3. There appears to be a bug when using the KaTeX rendering when it falls back to MathJax:

   <img width="626" alt="Screen Shot 2022-10-15 at 11 44 15 AM" src="https://user-images.githubusercontent.com/1028433/196003001-7d4f6a55-963b-48be-a00b-60cbd76ed941.png">

   Here my `\newcommand{\mat}[1]{\mathbf{#1}}` from the previous cell is apparently ignored.  I have filed an issue:

   *  https://github.com/sagemathinc/cocalc/issues/6179

**Comment by realfogman** [comment]
Hello mforbes, thank you very much for pointing me in the right direction.
I switched to the classical jupyter notebook and the \newcommands worked across the cells.
The only thing, which I'm missing is the autonumbering of equations across cells.
There exist a plugin through nbextentions, which I guess cannot be easily incorporated into cocalc?

**Comment by mforbes** [comment]
I will let someone from CoCalc address this, but I think many of these features work in the CoCalc interface... but there might still be some outstanding bugs.  The classic server always works, but then you lose collaboration.

**Comment by williamstein** [comment]
My understanding is the root cause of your issue was \DeclareMathOperator not existing in katex.  I've implemented it, so it's now available and live.
